sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,566 @@
+            modules: List of modules associated with the distribution object.
+
+        """
+        self._q_arg = q
+        if isinstance(q, Distribution):
+            q = adapt_variational_distribution(
+                q,
+                self._prior,
+                self.link_transform,
+                parameters=parameters,
+                modules=modules,
+            )
+            make_object_deepcopy_compatible(q)
+            self_custom_q_init_cache = deepcopy(q)
+            self._q_build_fn = lambda *args, **kwargs: self_custom_q_init_cache
+            self._trained_on = None
+        elif isinstance(q, str) or isinstance(q, Callable):
+            if isinstance(q, str):
+                self._q_build_fn = get_flow_builder(q)
+            else:
+                self._q_build_fn = q
+
+            q = self._q_build_fn(
+                self._prior.event_shape,
+                self.link_transform,
+                device=self._device,
+            )
+            make_object_deepcopy_compatible(q)
+            self._trained_on = None
+        elif isinstance(q, VIPosterior):
+            self._q_build_fn = q._q_build_fn
+            self._trained_on = q._trained_on
+            self.vi_method = q.vi_method
+            self._device = q._device
+            self._prior = q._prior
+            self._x = q._x
+            self._q_arg = q._q_arg
+            make_object_deepcopy_compatible(q.q)
+            q = deepcopy(q.q)
+        move_all_tensor_to_device(q, self._device)
+        check_variational_distribution(q, self._prior)
+        self._q = q
+
+    @property
+    def vi_method(self) -> str:
+        """Variational inference method e.g. one of [rKL, fKL, IW, alpha]."""
+        return self._vi_method
+
+    @vi_method.setter
+    def vi_method(self, method: str) -> None:
+        """See `set_vi_method`."""
+        self.set_vi_method(method)
+
+    def set_vi_method(self, method: str) -> "VIPosterior":
+        """Sets variational inference method.
+
+        Args:
+            method: One of [rKL, fKL, IW, alpha].
+
+        Returns:
+            `VIPosterior` for chainable calls.
+        """
+        self._vi_method = method
+        self._optimizer_builder = get_VI_method(method)
+        return self
+
     def sample(
         self,
         sample_shape: Shape = torch.Size(),
-        show_progress_bars: bool = True,
-    ):
-        raise NotImplementedError
+        x: Optional[Tensor] = None,
+        method: str = "naive",
+        **kwargs,
+    ) -> Tensor:
+        """Samples from the variational posterior distribution.
+
+        Args:
+            sample_shape: Shape of samples
+            method: Sampling method, alternatively we can debias the approximation by
+                using simple and efficient sampling schemes. We support one of [naive,
+                sir].
+            kwargs: Hyperparameters for the sampling methods.
```

Comment:
rendering error here? naive and sir are described after the kwargs.

Reply:
Mhh, for all arguments that fall in the kwargs, I made an extra tabular:

main_args: doc
kwargs: Hyperparameters, which vary for different methods
     params: desc

This may cause problems with autodoc. Should I just expand the kwargs and indicate in the description something like "this is only used if method=sir"?

Reply:
Ahhhh, ok, I misunderstood. Sorry, all good then.